README.md
# rs3clans.py
[![PyPI](https://img.shields.io/pypi/v/rs3clans.svg)](https://pypi.org/project/rs3clans/) [![PyPI - Python Version](https://img.shields.io/pypi/pyversions/rs3clans.svg)](https://pypi.org/project/rs3clans/) [![Build Status](https://travis-ci.org/johnvictorfs/rs3clans.py.svg?branch=master)](https://travis-ci.org/johnvictorfs/rs3clans.py)

A Python 3 module wrapper for RuneScape 3 API

- [Change Log](CHANGELOG.md)

***
## Requirements:

- [`Python 3.6+`](https://www.python.org/)

- [`requests>=2.21.0`](http://docs.python-requests.org/en/master/)

***

## Setup with pip:

```bash
$ python3 -m pip install rs3clans
```

***

## Usage:

<details>
<summary>
Players
</summary>

- Creating a Player object
    - Always check if a player actually exists before doing anything with it
```python
>>> from rs3clans import players
>>> player = players.Player(name='nriver')
>>> if player.exists:
...     pass
```

- Whether the player exists or not
```python
>>> player.exists
True
```

- Whether his Runemetrics Profile is Private or not
```python
>>> player.private_profile
False
```

- You can also pass the argument runemetrics as `False` if you don't want their runemetrics info to be set
    - This will make you unable to use some attributes from the Player class
```python
>>> player = players.Player(name='nriver', runemetrics=False)
```

- Getting a player's name
    - (if his Runemetrics Profile is private it will return the same name passed when creating object)
```python
>>> player.name
'NRiver'
```

- Getting a player's total Exp (requires Public Runemetrics Profile)
```python
>>> player.exp
1037291112
```

- Getting a player's Total Level (requires Public Runemetrics Profile)
```python
>>> player.total_level
```

- Getting a player's Combat Level (requires Public Runemetrics Profile)
```python
>>> player.combat_level
138
```

- Quests information about a player (requires Public Runemetrics Profile)
```python
>>> player.quests_not_started
32
>>> player.quests_started
5
>>> player.quests_complete
198
```

- Getting information on a specific skill of the player (requires Public Runemetrics Profile)
```python
>>> player.skill('agility').level
99
```

- Skill name is case-insensitive
```python
>>> player.skill('AtTaCk').rank
68311
```

- Can pass skill names as well as id
    - (8 = Woodcutting for example)
```python
>>> player.skill(8).exp
14054178.6
```

- Getting a player's title
```python
>>> player.title
'The Liberator'
```

- Verifying if a player's title is a suffix or not
```python
>>> player.suffix
True
```

- Getting a player's clan
```python
>>> player.clan
'Atlantis'
```
</details>

<details>
<summary>
Clans
</summary>

- Creating a Clan object
    - Always check if a clan actually exists before doing anything with it
```python
>>> from rs3clans import clans
>>> try:
...     clan = clans.Clan('Atlantis')
... except clans.ClanNotFoundError:
...     print('Clan not found.')
```

- Getting a clan's total Exp
```python
>>> clan.exp
151349638333
```

- Getting information about a specific member in that clan
    - Clan.member attribute (dict) (requires case-sensitive name)
    - Clan.get_member() (method) (does not require case-sensitive name)
    - Returns a ClanMember Object
```python
>>> # Case-sensitive
>>> clan.member['NRiver']
ClanMember(NRiver, Overseer, 1043065027)
>>> clan.member['NRiver'].rank
'Overseer'
```

```python
>>> # Case-insensitive
>>> clan.get_member('nriver')
ClanMember(NRiver, Overseer, 1043065027)
>>> clan.get_member('nRiVeR').rank
'Overseer'
```

- Getting the number of players in a clan
```python
>>> clan.count
499
```

- Getting the average Clan Exp per player in clan
```python
>>> clan.avg_exp
303305888.44288576
```

- Iterate through a Clan
```python
>>> for member in clan:
>>>     print(f"{member} - {member.name}")
ClanMember(Pedim, Owner, 1249520826) - Pedim
ClanMember(Acriano, Overseer, 1903276564) - Acriano
ClanMember(Cogu, Overseer, 1829449412) - Cogu
ClanMember(Black bullet, Overseer, 1100767386) - Black Bullet
ClanMember(NRiver, Overseer, 1090093362) - NRiver
ClanMember(Kurenaii, Overseer, 395850997) - Kurenaii
...
```

</details>

***

## Contributing:

- Guidelines:
    - Follow Pep8 (linter: flake8)

- [Fork](https://github.com/johnvictorfs/rs3clans.py/fork) the repository

- Install Dev dependencies
    - With poetry
        ```bash
        $ python3 -m venv .venv
        $ poetry shell
        $ poetry install
        ```
    - Without poetry
        ```bash
        $ python3 -m venv .venv
        $ source .venv/bin/activate
        $ pip install -r requirements-dev.txt
        ```

- Run the tests to make sure everything is ok
    ```bash
    $ pytest
    ```

- Make your changes to the code in the directory `rs3clans.py/rs3clans/`

- Add necessary tests to the `rs3clans.py/tests/` directory ([`pytest docs`](https://docs.pytest.org/en/latest/))

- Run the tests again
    ```bash
    $ pytest
    ```

- Also run `flake8` just to check if the code style is also fine
    ```bash
    $ flake8 --ignore=E501 rs3clans/
    ```

- If everything went fine then send a [Pull Request](https://github.com/johnvictorfs/rs3clans.py/pulls)

***


CHANGELOG.md
# rs3clans.py Change Log

[![PyPI](https://img.shields.io/pypi/v/rs3clans.svg)](https://pypi.org/project/rs3clans/) [![PyPI - Python Version](https://img.shields.io/pypi/pyversions/rs3clans.svg)](https://pypi.org/project/rs3clans/) [![Build Status](https://travis-ci.org/johnvictorfs/rs3clans.py.svg?branch=master)](https://travis-ci.org/johnvictorfs/rs3clans.py)

***

### 1.1.3

- Fixed issues with usernames that had dashes (`-`) in their username being marked as non-existent players

***

### 1.1.1 and 1.1.2

- Removed random print statement when creating a `Clan` object

- Fixed issue with non-breaking spaces not being removed from Player names in a `Clan` object member attribute, meaning you couldn't use `Clan.get_member` or `Clan.member` to get a Player that had spaces in its name
    - Added tests to make sure this doesn't break again in the future

***

### 1.1.0

- Created new `ClanMember` class, that represents all Members in a Clan now, instead of a dictionary:
    ```python3
    >>> for member in clan:
    >>>     print(f"{member} - {member.name}")
    ClanMember(Pedim, Owner, 1249520826) - Pedim
    ClanMember(Acriano, Overseer, 1903276564) - Acriano
    ClanMember(Cogu, Overseer, 1829449412) - Cogu
    ClanMember(Black bullet, Overseer, 1100767386) - Black Bullet
    ClanMember(NRiver, Overseer, 1090093362) - NRiver
    ClanMember(Kurenaii, Overseer, 395850997) - Kurenaii
    ...
    ```
    
    `ClanMember.name` (str): Name of the Clan Member
    `ClanMember.rank` (str): Rank of the Clan Member
    `ClanMember.exp` (int): Clan Exp of the Clan Member

- Ditched Pipenv for [Poetry](https://github.com/sdispater/poetry/)

- Kept `requirements.txt` and `requirements-dev.txt` for usage without poetry

- No longer using `setup.py` for deploying since Poetry does it automatically

- Added checking for `flake8` in Travis-ci build

- Made all `clans` tests not do any request, soon to be done with the `players` module as well

- Updated All Doc Strings to Google Style, preparing to go back to auto-generated docs

- `Player`, `Clan` and `ClanMember` are now dataclasses

- `dataclasses` is now a dependency in Python < 3.7

***

### 1.0.8

- Fixed issue with not being able to create a `Player` on some specific cases if `runemetrics` argument was passed as `True` and its name had spaces on it

***

### 1.0.7

- Passing a `None` type value to `Clan` and `Player`'s `name` argument now raises `TypeError` as it should

***

### 1.0.6

- Changed all `__str__` methods to just `__repr__` methods that are more concise.
```python3
>>> # Old
>>> import rs3clans
>>> print(rs3clans.Player('nriver')
>>> Name: NRiver Clan: Atlantis Exists: True

>>> # New
>>> import rs3clans
>>> rs3clans.Player('nriver')
>>> Player(name='NRiver', clan='Atlantis', exists=True)
```

***

### 1.0.5
- `Clan` objects can now be iterated through as if you were iterating over `Clan.member.items()` (requires `set_dict` parameter when creating `Clan` object to be `True` (`True` by default))
    - Example:
        ```python3
        >>> clan = rs3clans.Clan('atlantis')
        >>> for member in clan:
        ...     print(member)
        ...
        ('Pedim', {'rank': 'Owner', 'exp': 962627411})
        ('Tusoroxo', {'rank': 'Deputy Owner', 'exp': 1157686923})
        ('Acriano', {'rank': 'Overseer', 'exp': 1857267476})
        ('Cogu', {'rank': 'Overseer', 'exp': 1576814120})
        ('Black bullet', {'rank': 'Overseer', 'exp': 909580894})
        ('NRiver', {'rank': 'Overseer', 'exp': 1043065027})
        (...)
        >>> clan = rs3clans.Clan('atlantis', set_dict=False)
        >>> for member in clan:
        ...     print(member)
        >>> 
        ```

- Now using requests for every `Player` method, instead of using urllib for one, no usage changes here

- Now the creation of a `Player` or `Clan` object raises `ConnectionError` if it isn't possible to connect to one or more of RS3's API's used
    - Also added the following attributes so the status codes of each request made in those objects can be diagnosed:
        - `Clan.hiscores_status_code`
        - `Player.runemetrics_status_code`
        - `Player.runemetrics_status_code`
    ```python3
    >>> import rs3clans
    >>> try:
    ...     clan = rs3clans.Clan('atlantis')
    ... except ConnectionError:
    ...     print("Error when connecting to RS3's Clan API")
    ...     print(f"Status Code: {clan.hiscores_status_code}")
    ... except rs3clans.ClanNotFoundError:
    ...     print(f"Error. Clan {clan.name} does not exist")
    ...
    >>> try:
    ...     player = rs3clans.Player('nriver', runemetrics=True)
    ... except ConnectionError:
    ...     print("Error when connecting to RS3's Player API")
    ...     print(f"Status Code from Runemetrics API: {player.runemetrics_status_code}")
    ...     print(f"Status Code from Player Details API: {player.runemetrics_status_code}")
    ...
    >>> 
    ```

***

### 1.0.4
- Fixed `Player.exists()` method always returning False unless `runemetrics` argument was passed as true or `Player.set_runemetrics_info()` was called

***

### 1.0.3
- `Clan.get_member(name)` method now returns the name of the player found in the dict as below:
    ```python3
    >>> from rs3clans import Clan
    >>> clan = Clan('atlantis')
    >>> clan.get_member('nriver')
    {'rank': 'Overseer', 'exp': 1043065027, 'name': 'NRiver'}
    ```

***

### 1.0.2
- Fixed `Clan.get_member(name)` not working with names that had spaces on them

***

### 1.0.1
- Added continuous integration with tests using [`Travis CI`](https://travis-ci.org)
    - [Project URL](https://travis-ci.org/johnvictorfs/rs3clans.py)

- Lots of changes to Documentation

***

### 1.0.0
- Added `runemetrics` (bool, default=True) argument to `Player`
    - If True, `Player.set_runemetrics_info()` will be ran by default

rs3clans/players.py
# Standard Imports
import json

# Non-standard Imports
import requests


class Player:
    """Class with attributes of a Runescape Player using Jagex's RS3 API

    Most of Runescape 3's API can be accessed at: http://runescape.wikia.com/wiki/Application_programming_interface

    ...

    Parameters
    ----------
    name : str
        Name of the Player, gets set to case-sensitive name later if his RuneMetrics profile is not private.
        If you're also working with clans and the player's profile is private, name should be passed case-sensitively.
        Otherwise some clan attributes will not be available.
    runemetrics : bool
        Where or not Runemetrics player info should be set or not, slows down creating the object, but provides more
        attributes and methods.

    Attributes
    ----------
    name : str
        Name of the Player, gets set to case-sensitive name later if his RuneMetrics profile is not private.
    metrics_info : dict
        Runemetrics Info in dictionary format about the player.
    exp : int or None
        Total Exp of the player, or None if private_profile is True.
    combat_level : int or None
        Total Combat Level of the player, or None if private_profile is True.
    total_level : int or None
        Total Skill Level of the player, or None if private_profile is True.
    quests_not_started : int
        Number of quests the player has not started.
    quests_started : int
        Number of the quests the player has started.
    quests_complete : int
        Number of quests the player has completed.
    private_profile : bool
        Whether the player has a private RuneMetrics profile or not.
    exists : bool
        Whether the player exists or not.
    details_status_code : int
        The status code of the GET request to RS3 Player Details API
    runemetrics_status_code : int
        The status code of the GET request to RS3 Player Runemetrics API

    Raises
    ------
    ConnectionError
        If connecting to any of RS3's API related to players fails.
    """

    def __init__(self, name: str, runemetrics: bool = True):

        self.name = name
        self.metrics_info = None
        self.exp = None
        self.combat_level = None
        self.total_level = None
        self.quests_not_started = None
        self.quests_started = None
        self.quests_complete = None
        self.skill_values = None
        self.private_profile = True
        self.exists = True

        self.details_status_code = None
        self.runemetrics_status_code = None

        self.info = self._dict_info()
        self.suffix = self.info['is_suffix']
        self.title = self.info['title']

        # TODO: The check below is a band-aid fix that marks self.exists as False even if 'runemetrics' parameter
        # is passed as False when creating Player object, a real way to check if a player exists needs to be added
        if not self._valid_username(self.name):
            self.exists = False

        if runemetrics:
            self.set_runemetrics_info()

        try:
            self.clan = self.info['clan']
        except KeyError:
            self.clan = None

    @staticmethod
    def _valid_username(username: str):
        """
        Checks if a username is in the valid character limit set by Jagex
        and if it doesn't have any invalid characters in it.

        Parameters
        ----------
        username : str
            The username to be checked if valid or not

        Returns
        -------
        bool
            True if the username is valid, False if it is not
        """
        if len(username) > 12:
            return False
        banned_chars = (
            '!', '@', '#', '$', '%', '¨', '&', '*', '(', ')', '-', '[', ']', ',', '.', '~', '^'
                                                                                            '{', '}', ':', ';', "'",
            '"', 'ç', 'Ç', '?', '/', '°', '|', '\\', '`', '=', '+', '¬'
                                                                    '¹', '²', '³', '£', '¢', 'ª', 'º')
        if any(char in username for char in banned_chars):
            return False
        return True

    def _raw_info(self):
        """
        Gets JsonP information on a player, in string format.

        Returns
        -------
        str
            JsonP information on player, converted to string.

        Raises
        ------
        ConnectionError
            If connecting to the RS3 Runemetrics profile API was not possible.

        ------
        .. note:: Has to be formatted correctly into Json or other formats to be worked with properly first.
        """

        info_url = (f"http://services.runescape.com/m=website-data/"
                    f"playerDetails.ws?names=%5B%22{self.name}%22%5D&callback=jQuery000000000000000_0000000000&_=0")
        response = requests.get(info_url)
        self.details_status_code = response.status_code
        if response.status_code != 200:
            raise ConnectionError(f'Not able to connect to RS3 playerDetails API. Status code: {response.status_code}')
        return str(response.content)

    def _dict_info(self):
        """
        Gets the raw string info from self._raw_info() and formats it into Dictionary format.:
        Used to set self.info.

        Returns
        -------
        dict
            Dictionary is formatted as follows::
                >>> player_info = {
                ... 'is_suffix': True,
                ... 'recruiting': True,
                ... 'name': 'nriver',
                ... 'clan': 'Atlantis',
                ... 'title': 'the Liberator'
                ... }
            is_suffix : bool
                If the player's title is a Suffix or not
            recruiting : bool
                If the player's clan is set as Recruiting or not
            name : str
                The player's name, passed as is when creating object Player
            clan : str
                The player's clan name
            title : str
                The player's current title
        """
        str_info = self._raw_info()
        info_list = []

        # str_info[36] = Start of json format in URL '{'
        for letter in str_info[36:]:
            info_list.append(letter)
            if letter == '}':
                break
        info_list = ''.join(info_list)
        info_dict = json.loads(info_list)
        info_dict['is_suffix'] = info_dict['isSuffix']
        del info_dict['isSuffix']
        return info_dict

    def set_runemetrics_info(self):
        """
        Sets Runemetrics info attributes for the Player.

        Attributes set
        --------------
        name : str
            Case-sensitive player username.
        exp : int
            Total Exp of the Player.
        combat_level : int
            Total Combat Level of the Player.
        total_level : int
            Total Skill Level of the Player.
        """
        # If user's runemetrics profile is private, self.name will be the same as passed when creating object.
        # Otherwise it will get the correct case-sensitive name from his runemetrics profile.
        # Some other info like Total exp, Combat level and Total level will be created as well.
        self.metrics_info = self.runemetrics_info()
        if self.metrics_info is not False:
            self.name = self.metrics_info['name']
            self.exp = self.metrics_info['totalxp']
            self.combat_level = self.metrics_info['combatlevel']
            self.total_level = self.metrics_info['totalskill']
            self.quests_not_started = self.metrics_info['questsnotstarted']
            self.quests_started = self.metrics_info['questsstarted']
            self.quests_complete = self.metrics_info['questscomplete']
            self.skill_values = self.metrics_info['skillvalues']

    def runemetrics_info(self):
        """
        Gets player info from Runemetrics' API.
        Sets `private_profile` to True if his Runemetrics profile is found to be Private.

        Format::
            >>> {
            ... "magic":16216354,
            ... "questsstarted":5,
            ... "totalskill":2715,
            ... "questscomplete":198,
            ... "questsnotstarted":32,
            ... "totalxp":1037291112,
            ... "ranged":84195157,
            ... "activities":[
            ...
            ... ],
            ... # "skillvalues" has all skills and they are marked with their respective ID's, only 3 were shown here
            ... "skillvalues":[
            ...    {
            ...       "level":120,
            ...       "xp":1857550415,
            ...       "rank":12014,
            ...       "id":26
            ...    },
            ...    {
            ...       "level":99,
            ...       "xp":1202360390,
            ...       "rank":10370,
            ...       "id":6
            ...    },
            ...    {
            ...       "level":99,
            ...       "xp":841951573,
            ...       "rank":26941,
            ...       "id":3
            ...    },
            ... ],
            ... "name":"NRiver",
            ... "rank":"36,708",
            ... "melee":527931306,
            ... "combatlevel":138,
            ... "loggedIn":"false"
            ... }

        Returns
        ------
        dict or bool
            Runemetrics player info if the player profile is public, or False if it is Private.

        Raises
        ------
        ConnectionError
            If connecting to the RS3 Runemetrics profile API was not possible.
        """
        user_name = self.name.replace(' ', '%20')
        info_url = f"https://apps.runescape.com/runemetrics/profile/profile?user={user_name}&activities=0"
        response = requests.get(info_url)
        self.runemetrics_status_code = response.status_code
        if response.status_code != 200:
            raise ConnectionError(
                f'Not able to connect to RS3 Runemetrics player profile API. Status code: {response.status_code}')
        json_info = response.json()
        try:
            if json_info['error'] == 'PROFILE_PRIVATE':
                self.private_profile = True
                return False
            if json_info['error'] == 'NO_PROFILE':
                self.private_profile = False
                self.exists = False
                return False
        except KeyError:
            self.private_profile = False
            return json_info

    def skill(self, skill: str or int):
        """
        Gets information on a specific skill from the Player

        Usage::
            >>> # Example of usage of the method skill()
            >>> player = Player('NRiver')
            >>> player.skill('agility').level
            99
            >>> # Can pass skill names as well as id
            >>> # (8 = Woodcutting for example)
            >>> player.skill(8).exp
            14054178.6
            >>> player.skill('AtTaCk').rank
            68311

        Parameters
        ----------
        skill : str or int
            Skill name or id to get info of

        Returns
        ----------
        Skill object
            Attributes of Skill
            --------------------
                exp (or 'xp') : Total exp the Player has in that skill

                level : The level the Player has in that skill

                rank : The rank the Player has in that skill
        """
        skills_index = {
            0: 'attack',
            1: 'defence',
            2: 'strength',
            3: 'constitution',
            4: 'ranged',
            5: 'prayer',
            6: 'magic',
            7: 'cooking',
            8: 'woodcutting',
            9: 'fletching',
            10: 'fishing',
            11: 'firemaking',
            12: 'crafting',
            13: 'smithing',
            14: 'mining',
            15: 'herblore',
            16: 'agility',
            17: 'thieving',
            18: 'slayer',
            19: 'farming',
            20: 'runecrafting',
            21: 'hunter',
            22: 'construction',
            23: 'summoning',
            24: 'dungeoneering',
            25: 'divination',
            26: 'invention',
        }

        try:
            skill = int(skill)
            for _skill in self.skill_values:
                if _skill['id'] == skill:
                    return Skill(name=skills_index[skill], skill=_skill)
        except ValueError:
            skill = skill.lower()
            for index, value in skills_index.items():
                if skill == value:
                    skill = index
                    for _skill in self.skill_values:
                        if _skill['id'] == skill:
                            return Skill(name=skills_index[skill], skill=_skill)

    def __str__(self):
        return f"Name: {self.name} Clan: {self.clan} Exists: {self.exists}"


class Skill:
    def __init__(self, name: str, skill: dict):
        self.name = name
        try:
            self.exp = skill['xp'] / 10
        except KeyError:
            self.exp = None
        self.xp = self.exp
        try:
            self.level = skill['level']
        except KeyError:
            self.level = None
        try:
            self.rank = skill['rank']
        except KeyError:
            self.rank = None
        try:
            self.id = skill['id']
        except KeyError:
            self.id = None

    def __str__(self):
        return self.name
